L96-python-setuptools-plugin.md

```diff
@@ -0,0 +1,107 @@
+use this mechanism and solely rely on [setup.cfg][declarative_config].
+
+Custom commands are not the only integration point that is available. `setuptools` also supports
+entrypoints that allows for packages to act as a plugin and provide custom behavior. This is
+[documented with examples][setuptools_entrypoint] of generating files based on revision control
+information. This could be used to generate source files based on `.proto` files.
+
+[Original Feature Request][original_feature_request]
+
+## Proposal
+
+Add an entrypoint to `grpc-tools` that would register as a `file_finer` hook with `setuptools`. To
+simplify the implementation without introducing an additional code path, the entrypoint will target
+the same `main()` function used by the `grpc_tools.protoc` CLI. A [tool table][tool_table] within the
+file `pyproject.toml` will be used to accept configuration parameters. This will add a dependency on
+tomli for now, but this will [not be needed][pep_680] starting in Python 3.11.
+
+In order to prevent unintended behavior, the first thing the functionality will do is check the
+configuration file to see if the section has been configured. If the file does not exist or the section
+is not configured, no additional steps will be taken. `setuptools` will be provided with an empty
+list to indicate that there are no additional files to include in the distribution.
+
+## Rationale
+
+A downside of this proposal is that it will introduce yet another way to get the Python modules that
+correspond to the proto files.
+
+The primary way generate the Python files is using the
+[`grpc_tools.protoc` command line interface][protoc_cli]. The advantage of this mechanism of this
+method is also its disadvantage. It is a completely standalone way to generate the source code for
+the python modules. While projects are allowed to decide how to work this generated files in the way
+that best suits the project, there is no integration point. This means the project **must** decide
+how and when this command is executed, likely needing to write a script to execute this CLI.
+
+For projects that use `setuptools` to publish a package, `grpc-tools` includes a
+[custom command][grpc_command]. However, it also requires a specific call like the `grpc_tools.protoc`
+CLI and is not compatible with installing from a source distribution for projects that declare
+`setuptools` as build system that should be used. Additionally, one of the top result for "grpc
+setuptools command" is [this post][setuptools_gen_question], which suggests a custom command that
+doesn't rely on the one provided by `grpc-tools`. This suggest that there is a desire for functionality
+that doesn't require a specific all and that the existing functionality is under-documented.
+
+gRPC also provides a way to get the Python [classes at runtime][runtime_classes]
+([Proposal][runtime_proposal]). However, it is still [marked as experimental][runtime_api] almost 2
```

Comment:
CC @gnossen I recall there are good adoptions to the runtime proto. Do you know if there is any blocker for making it stable?

Reply:
Graduating sounds good.